subiquity/controllers/installprogress.py

```diff
@@ -40,10 +49,109 @@ class InstallState:
     ERROR = -1
 
 
+class WaitForCurtinEventsTask(BackgroundTask):
+
+    def __init__(self, controller):
+        self.controller = controller
+        self.waited = 0.0
+
+    def start(self):
+        pass
+
+    def _bg_run(self):
+        while self.controller._event_indent and self.waited < 5.0:
+            time.sleep(0.1)
+            self.waited += 0.1
+        log.debug("waited %s seconds for events to drain", self.waited)
+
+    def end(self, observer, fut):
+        # Will raise if command failed:
+        fut.result()
+        self.controller._install_event_start("installing snaps")
+        observer.task_succeeded()
+
+
+class DownloadSnapTask(BackgroundTask):
+
+    def __init__(self, controller, download_dir, snap_name, channel):
+        self.controller = controller
+        self.download_dir = os.path.join(download_dir, snap_name)
+        self.snap_name = snap_name
+        self.channel = channel
+
+    def start(self):
+        self.controller._install_event_start(_("downloading {}").format(self.snap_name))
+        os.mkdir(self.download_dir)
+        self.proc = utils.start_command(
+            ['snap', 'download', '--channel='+self.channel, self.snap_name],
+            cwd=self.download_dir)
+
+    def _bg_run(self):
+        self.proc.communicate()
+        if self.proc.returncode != 0:
+            raise subprocess.CalledProcessError(
+                cp.returncode, cp.args, output=cp.stdout, stderr=cp.stderr)
+
+    def end(self, observer, fut):
+        self.controller._install_event_finish()
+        fut.result()
+        observer.task_succeeded()
+
+
+class UpdateSnapSeed(BackgroundTask):
+
+    def __init__(self, controller, root):
+        self.controller = controller
+        self._seed_yaml = os.path.join(root, "var/lib/snapd/seed/seed.yaml")
+        self._tmp_dir = os.path.join(root, "var/lib/snapd/seed/tmp")
+        self._snap_dir = os.path.join(root, "var/lib/snapd/seed/snaps")
+        self._assertions_dir = os.path.join(root, "var/lib/snapd/seed/assertions")
+
+    def start(self):
+        self.controller._install_event_start(_("updating snap seed"))
+
+    def _bg_run(self):
+        # This doesn't really need to be in the background, but as we
+        # have the infrastructure already in place, we may as well.
+        with open(self._seed_yaml) as fp:
+            seed = yaml.safe_load(fp)
+
+        for snap_name in os.listdir(self._tmp_dir):
+            [snap_path] = glob.glob(os.path.join(self._tmp_dir, snap_name, "*.snap"))
+            [assertion_path] = glob.glob(os.path.join(self._tmp_dir, snap_name, "*.assert"))
+
+            snap_file = os.path.basename(snap_path)
+            assertion_file = os.path.basename(assertion_path)
+            os.rename(snap_path, os.path.join(self._snap_dir, snap_file))
+            os.rename(assertion_path, os.path.join(self._assertions_dir, assertion_file))
+
+            os.rmdir(os.path.join(self._tmp_dir, snap_name))
```

Comment:
shutil.move() handles the rename and cleanup.

Reply:
I don't think it makes things any clearer here?

Reply:
Really? I trust shutil.move() to do this atomically using context managers. 
Don't we also get to skip the path joining?

shutil.move("path to downloaded snap", "target_dir")
shutil.move("path to downloaded assertion", "target_dir")

Reply:
Clearly I don't understand what you are suggesting here.